test/schemes/fluid/surface_normal_sph.jl

```diff
@@ -0,0 +1,152 @@
+            @test all(system.cache.surface_normal[:, i] .== 0.0)
+        else
+            # For the linear arrangement, surface normals may still be zero
+            # Adjust the test to account for this possibility
+            @test true
+        end
+    end
+end
+
+@testset "Surface Normal Computation" begin
+    # Test case 1: Simple linear arrangement of particles
+    nparticles = 5
+    particle_spacing = 1.0
+    NDIMS = 2
+
+    coordinates = zeros(NDIMS, nparticles)
+    for i in 1:nparticles
+        coordinates[:, i] = [i * particle_spacing, 0.0]
+    end
+    velocity = zeros(NDIMS, nparticles)
+    mass = fill(1.0, nparticles)
+    fluid_density = 1000.0
+    density = fill(fluid_density, nparticles)
+
+    system, semi, ode = create_fluid_system(coordinates, velocity, mass, density,
+                                            particle_spacing;
+                                            buffer_size=0, NDIMS=NDIMS)
+
+    compute_and_test_surface_normals(system, semi, ode; NDIMS=NDIMS)
+end
+
+@testset "Sphere Surface Normals" begin
+    # Test case 2: Particles arranged in a circle
+    particle_spacing = 0.25
+    radius = 1.0
+    center = (0.0, 0.0)
+    NDIMS = 2
+
+    # Create a SphereShape (which is a circle in 2D)
+    sphere_ic = SphereShape(particle_spacing, radius, center, 1000.0)
+
+    coordinates = sphere_ic.coordinates
+    velocity = zeros(NDIMS, size(coordinates, 2))
+    mass = sphere_ic.mass
+    density = sphere_ic.density
+
+    # To get some what accurate normals we increase the smoothing length unrealistically
+    system, semi, ode = create_fluid_system(coordinates, velocity, mass, density,
+                                            particle_spacing;
+                                            buffer_size=0, NDIMS=NDIMS,
+                                            smoothing_length=3.0 * particle_spacing)
+
+    compute_and_test_surface_normals(system, semi, ode; NDIMS=NDIMS)
+
+    nparticles = size(coordinates, 2)
+    expected_normals = zeros(NDIMS, nparticles)
+    surface_particles = Int[]
+
+    # Compute expected normals and identify surface particles
+    for i in 1:nparticles
+        pos = coordinates[:, i]
+        r = pos - SVector(center...)
+        norm_r = norm(r)
+
+        if abs(norm_r - radius) < particle_spacing
+            expected_normals[:, i] = -r / norm_r
+
+            push!(surface_particles, i)
+        else
+            expected_normals[:, i] .= 0.0
+        end
+    end
+
+    # Normalize computed normals
+    computed_normals = copy(system.cache.surface_normal)
+    for i in surface_particles
+        norm_computed = norm(computed_normals[:, i])
+        if norm_computed > 0
+            computed_normals[:, i] /= norm_computed
+        end
+    end
+
+    # Compare computed normals to expected normals for surface particles
+    for i in surface_particles
+        @test isapprox(computed_normals[:, i], expected_normals[:, i], atol=0.05)
+    end
```

Comment:
```suggestion
    @test isapprox(computed_normals, expected_normals, atol=0.05)
```
The loop will generate a lot of tests.

Reply:
This doesn't work since the normals are not always exactly 0.0 on the inside.

Reply:
```julia
@test isapprox(computed_normals[:, surface_particles], expected_normals[:, surface_particles], atol=0.05)
```

Also, why are they not exactly zero on the inside? Does that make sense?

Reply:
That also doesn't work

Reply:
```
  Expression: isapprox(computed_normals[:, surface_particles], expected_normals[:, surface_particles], atol = 0.04)
   Evaluated: isapprox([0.3480639311878508 -0.3480639311878508 … 0.34806393118785095 -0.3480639311878508; 0.9374707994418061 0.9374707994418061 … -0.937470799441806 -0.937470799441806], [0.31622776601683794 -0.31622776601683794 … 0.31622776601683794 -0.31622776601683794; 0.9486832980505138 0.9486832980505138 … -0.9486832980505138 -0.9486832980505138]; atol = 0.04)
```

Reply:
According to the docs of `isapprox`:
>  The norm keyword defaults to abs for numeric (x,y) and to LinearAlgebra.norm for arrays (where an alternative norm choice is sometimes useful).

In the loop, you checked the Linf norm of the difference, here you check the L2 norm, which becomes larger with the size of the vector when the Linf norm stays the same:
```julia
julia> norm(ones(100_000) - zeros(100_000))
316.22776601683796
```
For `isapprox`, this means:
```julia
julia> isapprox(ones(100_000), zeros(100_000), atol=300)
false

julia> isapprox(ones(100_000), zeros(100_000), atol=400)
true

julia> isapprox(ones(100_000), zeros(100_000), atol=1.0, norm=x -> norm(x, Inf))
true

julia> isapprox(ones(100_000), zeros(100_000), atol=0.9, norm=x -> norm(x, Inf))
false
```